libs/tlv/src/coder.rs

```diff
@@ -0,0 +1,165 @@
+use std::io::{Read, Write};
+
+use crate::length::Length;
+
+/// Encodes a value into a writer.
+pub trait Encode {
+    /// Converts the value into a byte representation and writes it to the
+    /// writer.
+    fn encode<W>(&self, writer: &mut W) -> std::io::Result<()>
+    where
+        W: Write;
```

Comment:
In an earlier version of this I had the generic parameter `W` at the trait level, but that made it harder to use in other places because any function that accepted an `Encode` also had to have the `W` as its own type parameter. This made, for example, implementing `tlv::to_vec` difficult or impossible.